packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,748 @@
+export interface Vis {
+  canvas?: HTMLCanvasElement;
+  colors: string[];
+  analyser?: AnalyserNode;
+  oscStyle?: "dots" | "solid" | "lines";
+  bandwidth?: "wide" | "thin";
+  coloring?: "fire" | "line" | "normal";
+  peaks?: boolean;
+  saFalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  saPeakFalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa?: "analyzer" | "oscilloscope" | "none";
+}
+import { FFT } from "./FFTNullsoft";
+import {
+  renderHeight,
+  smallVis,
+  PIXEL_DENSITY,
+  doubled,
+  isMWOpen,
+} from "./Vis";
+
+let saPeaks = new Int16Array(76).fill(0);
+let saData2 = new Float32Array(76).fill(0);
+let saData = new Int16Array(76).fill(0);
+let saFalloff = new Float32Array(76).fill(0);
+let sample = new Float32Array(76).fill(0);
+let barPeak = new Int16Array(76).fill(0); // Needs to be specified as Int16 else the peaks don't behave as they should
+let chunk: number;
+let uVar12: number;
+let falloff: number;
+let peakFalloff: number;
+
+let pushDown: number = 0;
```

Comment:
Having these mutable (values that get written to) values in the module scope instead of them being properties of a class, means that every visualizer on the page will share access to the same values. So, if you have two Winamp's on the same page, they'll both be writing the same value and they'll corrupt each other's results.

Probably want you want is to either make them local variables declared within method (if they are only referenced in a single method) or defined as properties of the class that get initialized in that classes's constructor.

Reply:
It's possible that some of these values are used only in one method, but that they are expensive to construct, so we don't want to create them from scratch in each frame paint. In that case, we could just make them properties of the class even though they only get used in a single method.